src/commonMain/kotlin/org/kosat/CDCL.kt

```diff
@@ -3,20 +3,19 @@ package org.kosat
 import kotlin.math.abs
 import kotlin.math.max
 
-// CDCL
-fun solveCnf(cnf: CnfRequest): List<Int>? {
-    val clauses = (cnf.clauses.map { it.lit }).toMutableList()
-    val solver = Kosat(clauses, cnf.vars)
-    return if (solver.solve()) solver.getModel() else null
-}
+abstract class CDCL(val clauses: MutableList<MutableList<Int>>, initNumber: Int = 0) {
+
+    abstract fun solve(): List<Int>?
+
+    protected abstract fun getNextVariable(level: Int): Int
 
-class CDCL(val clauses: MutableList<MutableList<Int>>, initNumber: Int = 0) {
     var varsNumber = initNumber
-        private set
+        protected set
```

Comment:
I think the invariant of `varsNumber` is too important to allow yourself to mutate it from everywhere (even if "everywhere" is only the inheritance tree (*its very existence is bad by itself, by this is another story, see above*)). Maybe add (final) method "newVar" – the only one who will be allowed to increment `varsNumber`?
Or maybe just make it `private`, but mutate only from `CDCL` – in this case you have to provide all corresponding methods (I think `newVar`, `reserveVars`, `reset` are the ones which must have exclusive access to `varsNumber`).

Also, maybe rename to `numberOfVariable` or `topVar`?